File: pa7-ir-v2/python/IRSystem.py
#!/usr/bin/env python
import json
import math
import os
import re
import sys

from PorterStemmer import PorterStemmer


class IRSystem:

    def __init__(self):
        # For holding the data - initialized in read_data()
        self.titles = []
        self.docs = []
        self.vocab = []
        # For the text pre-processing.
        self.alphanum = re.compile('[^a-zA-Z0-9]')
        self.p = PorterStemmer()


    def get_uniq_words(self):
        uniq = set()
        for doc in self.docs:
            for word in doc:
                uniq.add(word)
        return uniq


    def __read_raw_data(self, dirname):
        print "Stemming Documents..."

        titles = []
        docs = []
        os.mkdir('%s/stemmed' % dirname)
        title_pattern = re.compile('(.*) \d+\.txt')

        # make sure we're only getting the files we actually want
        filenames = []
        for filename in os.listdir('%s/raw' % dirname):
            if filename.endswith(".txt") and not filename.startswith("."):
                filenames.append(filename)

        for i, filename in enumerate(filenames):
            title = title_pattern.search(filename).group(1)
            print "    Doc %d of %d: %s" % (i+1, len(filenames), title)
            titles.append(title)
            contents = []
            f = open('%s/raw/%s' % (dirname, filename), 'r')
            of = open('%s/stemmed/%s.txt' % (dirname, title), 'w')
            for line in f:
                # make sure everything is lower case
                line = line.lower()
                # split on whitespace
                line = [xx.strip() for xx in line.split()]
                # remove non alphanumeric characters
                line = [self.alphanum.sub('', xx) for xx in line]
                # remove any words that are now empty
                line = [xx for xx in line if xx != '']
                # stem words
                line = [self.p.stem(xx) for xx in line]
                # add to the document's conents
                contents.extend(line)
                if len(line) > 0:
                    of.write(" ".join(line))
                    of.write('\n')
            f.close()
            of.close()
            docs.append(contents)
        return titles, docs


    def __read_stemmed_data(self, dirname):
        print "Already stemmed!"
        titles = []
        docs = []

        # make sure we're only getting the files we actually want
        filenames = []
        for filename in os.listdir('%s/stemmed' % dirname):
            if filename.endswith(".txt") and not filename.startswith("."):
                filenames.append(filename)

        if len(filenames) != 60:
            msg = "There are not 60 documents in ../data/RiderHaggard/stemmed/\n"
            msg += "Remove ../data/RiderHaggard/stemmed/ directory and re-run."
            raise Exception(msg)

        for i, filename in enumerate(filenames):
            title = filename.split('.')[0]
            titles.append(title)
            contents = []
            f = open('%s/stemmed/%s' % (dirname, filename), 'r')
            for line in f:
                # split on whitespace
                line = [xx.strip() for xx in line.split()]
                # add to the document's conents
                contents.extend(line)
            f.close()
            docs.append(contents)

        return titles, docs


    def read_data(self, dirname):
        """
        Given the location of the 'data' directory, reads in the documents to
        be indexed.
        """
        # NOTE: We cache stemmed documents for speed
        #       (i.e. write to files in new 'stemmed/' dir).

        print "Reading in documents..."
        # dict mapping file names to list of "words" (tokens)
        filenames = os.listdir(dirname)
        subdirs = os.listdir(dirname)
        if 'stemmed' in subdirs:
            titles, docs = self.__read_stemmed_data(dirname)
        else:
            titles, docs = self.__read_raw_data(dirname)

        # Sort document alphabetically by title to ensure we have the proper
        # document indices when referring to them.
        ordering = [idx for idx, title in sorted(enumerate(titles),
            key = lambda xx : xx[1])]

        self.titles = []
        self.docs = []
        numdocs = len(docs)
        for d in range(numdocs):
            self.titles.append(titles[ordering[d]])
            self.docs.append(docs[ordering[d]])

        # Get the vocabulary.
        self.vocab = [xx for xx in self.get_uniq_words()]


    def compute_tfidf(self):
        # -------------------------------------------------------------------
        # TODO: Compute and store TF-IDF values for words and documents.
        #       Recall that you can make use of:
        #         * self.vocab: a list of all distinct (stemmed) words
        #         * self.docs: a list of lists, where the i-th document is
        #                   self.docs[i] => ['word1', 'word2', ..., 'wordN']
        #       NOTE that you probably do *not* want to store a value for every
        #       word-document pair, but rather just for those pairs where a
        #       word actually occurs in the document.
        print "Calculating tf-idf..."
        self.tfidf = {}
        for word in self.vocab:
            for d in range(len(self.docs)):
                if word not in self.tfidf:
                    self.tfidf[word] = {}
                self.tfidf[word][d] = 0.0

        # ------------------------------------------------------------------


    def get_tfidf(self, word, document):
        # ------------------------------------------------------------------
        # TODO: Return the tf-idf weigthing for the given word (string) and
        #       document index.
        tfidf = 0.0
        # ------------------------------------------------------------------
        return tfidf


    def get_tfidf_unstemmed(self, word, document):
        """
        This function gets the TF-IDF of an *unstemmed* word in a document.
        Stems the word and then calls get_tfidf. You should *not* need to
        change this interface, but it is necessary for submission.
        """
        word = self.p.stem(word)
        return self.get_tfidf(word, document)


    def index(self):
        """
        Build an index of the documents.
        """
        print "Indexing..."
        # ------------------------------------------------------------------
        # TODO: Create an inverted index.
        #       Granted this may not be a linked list as in a proper
        #       implementation.
        #       Some helpful instance variables:
        #         * self.docs = List of documents
        #         * self.titles = List of titles

        inv_index = {}
        for word in self.vocab:
            inv_index[word] = []

        self.inv_index = inv_index

        # ------------------------------------------------------------------


    def get_posting(self, word):
        """
        Given a word, this returns the list of document indices (sorted) in
        which the word occurs.
        """
        # ------------------------------------------------------------------
        # TODO: return the list of postings for a word.
        posting = []

        return posting
        # ------------------------------------------------------------------


    def get_posting_unstemmed(self, word):
        """
        Given a word, this *stems* the word and then calls get_posting on the
        stemmed word to get its postings list. You should *not* need to change
        this function. It is needed for submission.
        """
        word = self.p.stem(word)
        return self.get_posting(word)


    def boolean_retrieve(self, query):
        """
        Given a query in the form of a list of *stemmed* words, this returns
        the list of documents in which *all* of those words occur (ie an AND
        query).
        Return an empty list if the query does not return any documents.
        """
        # ------------------------------------------------------------------
        # TODO: Implement Boolean retrieval. You will want to use your
        #       inverted index that you created in index().
        # Right now this just returns all the possible documents!
        docs = []
        for d in range(len(self.docs)):
            docs.append(d)

        # ------------------------------------------------------------------

        return sorted(docs)   # sorted doesn't actually matter


    def rank_retrieve(self, query):
        """
        Given a query (a list of words), return a rank-ordered list of
        documents (by ID) and score for the query.
        """
        scores = [0.0 for xx in range(len(self.docs))]
        # ------------------------------------------------------------------
        # TODO: Implement cosine similarity between a document and a list of
        #       query words.

        # Right now, this code simply gets the score by taking the Jaccard
        # similarity between the query and every document.
        words_in_query = set()
        for word in query:
            words_in_query.add(word)

        for d, doc in enumerate(self.docs):
            words_in_doc = set(doc)
            scores[d] = len(words_in_query.intersection(words_in_doc)) \
                    / float(len(words_in_query.union(words_in_doc)))

        # ------------------------------------------------------------------

        ranking = [idx for idx, sim in sorted(enumerate(scores),
            key = lambda xx : xx[1], reverse = True)]
        results = []
        for i in range(10):
            results.append((ranking[i], scores[ranking[i]]))
        return results


    def process_query(self, query_str):
        """
        Given a query string, process it and return the list of lowercase,
        alphanumeric, stemmed words in the string.
        """
        # make sure everything is lower case
        query = query_str.lower()
        # split on whitespace
        query = query.split()
        # remove non alphanumeric characters
        query = [self.alphanum.sub('', xx) for xx in query]
        # stem words
        query = [self.p.stem(xx) for xx in query]
        return query


    def query_retrieve(self, query_str):
        """
        Given a string, process and then return the list of matching documents
        found by boolean_retrieve().
        """
        query = self.process_query(query_str)
        return self.boolean_retrieve(query)


    def query_rank(self, query_str):
        """
        Given a string, process and then return the list of the top matching
        documents, rank-ordered.
        """
        query = self.process_query(query_str)
        return self.rank_retrieve(query)


def run_tests(irsys):
    print "===== Running tests ====="

    ff = open('../data/queries.txt')
    questions = [xx.strip() for xx in ff.readlines()]
    ff.close()
    ff = open('../data/solutions.txt')
    solutions = [xx.strip() for xx in ff.readlines()]
    ff.close()

    epsilon = 1e-4
    for part in range(4):
        points = 0
        num_correct = 0
        num_total = 0

        prob = questions[part]
        soln = json.loads(solutions[part])

        if part == 0:     # inverted index test
            print "Inverted Index Test"
            words = prob.split(", ")
            for i, word in enumerate(words):
                num_total += 1
                posting = irsys.get_posting_unstemmed(word)
                if set(posting) == set(soln[i]):
                    num_correct += 1

        elif part == 1:   # boolean retrieval test
            print "Boolean Retrieval Test"
            queries = prob.split(", ")
            for i, query in enumerate(queries):
                num_total += 1
                guess = irsys.query_retrieve(query)
                if set(guess) == set(soln[i]):
                    num_correct += 1

        elif part == 2:   # tfidf test
            print "TF-IDF Test"
            queries = prob.split("; ")
            queries = [xx.split(", ") for xx in queries]
            queries = [(xx[0], int(xx[1])) for xx in queries]
            for i, (word, doc) in enumerate(queries):
                num_total += 1
                guess = irsys.get_tfidf_unstemmed(word, doc)
                if guess >= float(soln[i]) - epsilon and \
                        guess <= float(soln[i]) + epsilon:
                    num_correct += 1

        elif part == 3:   # cosine similarity test
            print "Cosine Similarity Test"
            queries = prob.split(", ")
            for i, query in enumerate(queries):
                num_total += 1
                ranked = irsys.query_rank(query)
                top_rank = ranked[0]
                if top_rank[0] == soln[i][0]:
                    if top_rank[1] >= float(soln[i][1]) - epsilon and \
                            top_rank[1] <= float(soln[i][1]) + epsilon:
                        num_correct += 1

        feedback = "%d/%d Correct. Accuracy: %f" % \
                (num_correct, num_total, float(num_correct)/num_total)
        if num_correct == num_total:
            points = 3
        elif num_correct > 0.75 * num_total:
            points = 2
        elif num_correct > 0:
            points = 1
        else:
            points = 0

        print "    Score: %d Feedback: %s" % (points, feedback)


def main(args):
    irsys = IRSystem()
    irsys.read_data('../data/RiderHaggard')
    irsys.index()
    irsys.compute_tfidf()

    if len(args) == 0:
        run_tests(irsys)
    else:
        query = " ".join(args)
        print "Best matching documents to '%s':" % query
        results = irsys.query_rank(query)
        for docId, score in results:
            print "%s: %e" % (irsys.titles[docId], score)


if __name__ == '__main__':
    args = sys.argv[1:]
    main(args)

File: pa7-ir-v2/java/IRSystem.java
import java.io.*;
import java.util.Arrays;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;


public class IRSystem {
  List<String> titles;
  ArrayList<ArrayList<String>> documents;
  ArrayList<String> vocab;
  HashMap<String, ArrayList<Integer>> invertedIndex;

  // For the text pre-processing.
  PorterStemmer stemmer;

  CounterMap<String, Integer> tfidf;  // word and document index
  Counter<String> docFreq;

  public IRSystem(String dataDir) {
    stemmer = new PorterStemmer();
    readData(dataDir);
  }

  void index() {
    System.out.println("Indexing...");
    /** TODO: Create an inverted index.
     *  Granted this may not be a linked list as in a proper
     *  implementation.
     *  Some helpful instance variables:
     *    * documents
     */


    invertedIndex = new HashMap<String, ArrayList<Integer>>();
    for (String word : vocab) {
      invertedIndex.put(word, new ArrayList<Integer>());
    }

    /* End TODO */
  }


  void computeTFIDF()
  {
    /** TODO: Compute and store TF-IDF values for words and documents.
     *  Recall that you can make use of the instance variables:
     *   * vocab
     *   * documents
     * NOTE that you probably do *not* want to store a value for every
     * word-document pair, but rather just for those pairs where a
     * word actually occurs in the document.
     */
    System.out.println("Computing TF-IDF...");

    tfidf = new CounterMap<String, Integer>();
    for (String word : vocab) {
      for (int d = 0; d < documents.size(); d++) {
        tfidf.setCount(word, new Integer(d), 0.0);
      }
    }

    /* End TODO */
  }


  double getTFIDF(String word, int doc) {
    double tfidf_score = 0.0;
    /** TODO: Return the tf-idf weigthing for the given word (string) and
     * document index.
     */


    /* End TODO */
    return tfidf_score;
  }


  public double getTFIDFUnstemmed(String word, int doc) {
    /**
     * This function gets the TF-IDF of an *unstemmed* word in a document.
     * Stems the word and then calls get_tfidf. You should *not* need to
     * change this interface, but it is necessary for submission.
     */
    word = stemmer.stem(word);
    return getTFIDF(word, doc);
  }


  ArrayList<Integer> getPosting(String word) {
    ArrayList<Integer> posting = new ArrayList<Integer>();
    /** TODO: Return the list of postings for a word. */


    /** End TODO */
    return posting;
  }


  public ArrayList<Integer> getPostingUnstemmed(String word) {
    /**
     * Given a word, this *stems* the word and then calls get_posting on the
     * stemmed word to get its postings list. You should *not* need to change
     * this function. It is needed for submission.
     */
    word = stemmer.stem(word);
    return getPosting(word);
  }


  ArrayList<Integer> booleanRetrieve(ArrayList<String> query)
  {
    /**
     * Given a query in the form of a list of *stemmed* words, this returns
     * the list of documents in which *all* of those words occur (ie an AND
     * query).
     */

    /** TODO: Implement Boolean retrieval. You will want to use your
     * inverted index that you created in index().
     * Right now this just returns all the possible documents!
     * Return an empty list if the query does not return any documents.
     */

    ArrayList<Integer> docs = new ArrayList<Integer>();
    for (int d = 0; d < documents.size(); d++) {
        docs.add(new Integer(d));
    }


    /* End TODO */
    Collections.sort(docs);
    return docs;
  }


  PriorityQueue<Integer> rankRetrieve(ArrayList<String> query) {
    double scores[] = new double[documents.size()];

    /*************************************************************/
    /** TODO: Implement cosine similarity. */
    /*  For now just implements Jaccard similarity. */
    HashSet<String> wordsInQuery = new HashSet<String>();
    wordsInQuery.addAll(query);
    HashSet<String> wordsInDoc;
    HashSet<String> setUnion;
    HashSet<String> setIntersection;
    for (int d = 0; d < documents.size(); d++) {
      wordsInDoc = new HashSet<String>();
      wordsInDoc.addAll(documents.get(d));
      setUnion = new HashSet<String>(wordsInDoc);
      setUnion.addAll(wordsInQuery);
      setIntersection = new HashSet<String>(wordsInQuery);
      setIntersection.retainAll(wordsInDoc);
      scores[d] = ((double) setIntersection.size()) / ((double) setUnion.size());
    }

    /* End TODO */

    PriorityQueue<Integer> pq = new PriorityQueue<Integer>();
    for (int d = 0; d < scores.length; d++) {
      pq.add(new Integer(d), scores[d]);
    }

    PriorityQueue<Integer> topTen = new PriorityQueue<Integer>();
    for (int d = 0; d < 10; d++) {
      double priority = pq.getPriority();
      topTen.add(pq.next(), priority);
    }

    return topTen;
  }


  /** Given a query string, processes the string and returns the list of
   * lowercase, alphanumeric, stemmed words in the string.
   */
  ArrayList<String> processQuery(String queryString)
  {
    // lowercase
    queryString = queryString.toLowerCase();
    ArrayList<String> query = new ArrayList<String>();
    for (String s : queryString.split("\\s+")) {
      // remove non alphanumeric characters
      s = s.replaceAll("[^a-zA-Z0-9]", "");
      // stem s
      s = stemmer.stem(s);
      if (!s.equals(""))
        query.add(s);
    }
    return query;
  }


  /** Given a string, this will process and then return the list of matching
   * documents found by booleanRetrieve()
   */
  ArrayList<Integer> queryRetrieve(String queryString)
  {
    ArrayList<String> query = processQuery(queryString);
    return booleanRetrieve(query);
  }


  /** Given a string, this will process and then return the list of the
   * top matching documents found by rankRetrieve()
   */
  PriorityQueue<Integer> queryRank(String queryString)
  {
    ArrayList<String> query = processQuery(queryString);
    return rankRetrieve(query);
  }


  void getUniqWords() {
    HashSet<String> uniqWords = new HashSet<String>();
    for (ArrayList<String> document : documents) {
      for (String word : document) {
        uniqWords.add(word);
      }
    }
    vocab = new ArrayList<String>(uniqWords);
  }


  ArrayList<String> readRawFile(String title, BufferedReader input, String stemmedDirName)
  {
    ArrayList<String> document = new ArrayList<String>();
    /* Output buffer for stemmed document. */
    String stemmedFile = stemmedDirName + "/" + title + ".txt";
    BufferedWriter output = null;
    try {
      output = new BufferedWriter(new FileWriter(new File(stemmedFile)));
    } catch(IOException e) {
      System.err.println("Error opening stemmed cache file for " + title);
      e.printStackTrace();
      System.exit(1);
    }

    try {
      String line;
      while ((line = input.readLine()) != null) {
        // make sure everything is lowercase
        line = line.toLowerCase();
        // split on whitespace
        ArrayList<String> tmp = new ArrayList<String>();
        boolean emptyLine = true;
        for (String s : line.split("\\s+")) {
          // Remove non alphanumeric characters
          s = s.replaceAll("[^a-zA-Z0-9]", "");
          // Stem word.
          s = stemmer.stem(s);

          if (!s.equals("")) {
            tmp.add(s);

            /* Write to stemmed file. */
            if (!emptyLine) {
              output.write(" ", 0, 1);
            }
            output.write(s, 0, s.length());
            emptyLine = false;
          }
        }

        /* Write new line to stemmed file (if you put anything on the line). */
        if (output != null && !emptyLine) {
          output.newLine();
        }

        document.addAll(tmp);
      }
    } catch(IOException e) {
      System.err.println("Error closing stemmed cache file for " + title);
      e.printStackTrace();
      System.exit(1);
    }

    /* Close the stemmed file. */
    try {
      output.close();
    } catch(IOException e) {
      System.err.println("Error closing stemmed: " + title);
      e.printStackTrace();
      System.exit(1);
    }

    return document;
  }

  ArrayList<String> readStemmedFile(String title, BufferedReader input)
  {
    ArrayList<String> document = new ArrayList<String>();
    try {
      String line;
      while ((line = input.readLine()) != null) {
        // make sure everything is lowercase
        line = line.toLowerCase();
        // split on whitespace
        ArrayList<String> tmp = new ArrayList<String>();
        for (String s : line.split("\\s+")) {
          tmp.add(s);
        }
        document.addAll(tmp);
      }
    } catch(IOException e) {
      System.err.println("Error processing document: " + title);
      e.printStackTrace();
      System.exit(1);
    }
    return document;
  }


  void readStemmedData(String dirName) {
    System.out.println("Already stemmed!");
    titles = new ArrayList<String>();
    documents = new ArrayList<ArrayList<String>>();
    ArrayList<File> files = new ArrayList<File>();
    for (File f : new File(dirName).listFiles()) {
      if (f.getName().endsWith(".txt") && !f.getName().startsWith("."))
        files.add(f);
    }
    if (files.size() != 60) {
      System.err.println("There are not 60 documents in ../data/RiderHaggard/stemmed\n"
          + "Remove ../data/RiderHaggard/stemmed directory and re-run.");
      System.exit(1);
    }
    for (File f : files) {
      try {
        BufferedReader input = new BufferedReader(new FileReader(f));
        String title = f.getName().replaceAll("\\.txt", "");
        titles.add(title);
        documents.add(readStemmedFile(title, input));
      } catch(IOException e) {
        e.printStackTrace();
        System.exit(1);
      }
    }
  }

  void readRawData(String dirName) {
    System.out.println("Stemming Documents...");
    String rawDirName = dirName + "/raw";
    String stemmedDirName = dirName + "/stemmed";
    new File(stemmedDirName).mkdir();
    titles = new ArrayList<String>();
    documents = new ArrayList<ArrayList<String>>();
    int i = 1;

    for (File f : new File(rawDirName).listFiles()) {
      if (f.getName().endsWith(".txt") && !f.getName().startsWith(".")) {
        System.out.println("Filename: " + f.getName());
        try {
          BufferedReader input = new BufferedReader(new FileReader(f));
          String title = f.getName().replaceAll(" \\d+\\.txt", "");
          titles.add(title);
          System.out.println("    Doc " + i + ": " + title);
          documents.add(readRawFile(title, input, stemmedDirName));
        } catch(IOException e) {
          e.printStackTrace();
          System.exit(1);
        }
        i++;
      }
    }
  }

  void readData(String dirName) {
    System.out.println("Reading in documents...");
    String[] subdirs = new File(dirName).list();
    boolean haveStemmed = false;
    for (String subdir : subdirs) {
      if (subdir.equals("stemmed")) {
        haveStemmed = true;
        break;
      }
    }
    if (haveStemmed) {
      readStemmedData(dirName + "/stemmed");
    } else {
      readRawData(dirName);
    }

    ArrayList<String> titlesSorted = new ArrayList<String>(titles);
    Collections.sort(titlesSorted);
    ArrayList<ArrayList<String>> documentsSorted = new ArrayList<ArrayList<String>>();
    for (int i = 0; i < titles.size(); i++) {
      String title = titlesSorted.get(i);
      documentsSorted.add(documents.get(titles.indexOf(title)));
    }

    documents = documentsSorted;
    titles = titlesSorted;

    getUniqWords();
  }


  String getTitle(int titleIndex) {
    return titles.get(titleIndex);
  }


  public static void runTests(IRSystem irSys) {
    ArrayList<String> questions = null;
    ArrayList<String> solutions = null;

    try {
      BufferedReader input = new BufferedReader(new FileReader(
            new File("../data/queries.txt")));
      questions = new ArrayList<String>();
      String line;
      while ((line = input.readLine()) != null) {
        questions.add(line);
      }
    } catch (IOException e) {
      System.err.println("Error reading ../data/queries.txt: "
          + e.getMessage());
    }
    try {
      BufferedReader input = new BufferedReader(new FileReader(
            new File("../data/solutions_java.txt")));
      solutions = new ArrayList<String>();
      String line;
      while ((line = input.readLine()) != null) {
        solutions.add(line);
      }
    } catch (IOException e) {
      System.err.println("Error reading ../data/solutions_java.txt: "
          + e.getMessage());
    }

    double epsilon = 1E-4;
    int numTests = solutions.size();
    for (int part = 0; part < numTests; part++) {

      int numCorrect = 0;
      int numTotal = 0;

      String problem = questions.get(part);
      String soln = solutions.get(part);

      if (part == 0) {    // Inverted Index test
        System.out.println("Inverted Index Test");

        String[] words = problem.split(", ");
        String[] golds = soln.split("; ");

        for (int i = 0; i < words.length; i++) {
          numTotal++;
          String word = words[i];
          HashSet<Integer> guess = new HashSet<Integer>(irSys.getPostingUnstemmed(word));
          String[] goldList = golds[i].split(", ");
          HashSet<Integer> goldSet = new HashSet<Integer>();
          for (String s : goldList) {
            goldSet.add(new Integer(s));
          }
          if (guess.equals(goldSet)) {
            numCorrect++;
          }
        }
      } else if (part == 1) {   // Boolean retrieval test
        System.out.println("Boolean Retrieval Test");

        String[] queries = problem.split(", ");
        String[] golds = soln.split("; ");
        for (int i = 0; i < queries.length; i++) {
          numTotal++;
          String query = queries[i];
          HashSet<Integer> guess = new HashSet<Integer>(irSys.queryRetrieve(query));
          String[] goldList = golds[i].split(", ");
          HashSet<Integer> goldSet = new HashSet<Integer>();
          for (String s : goldList) {
            goldSet.add(new Integer(s));
          }
          if (guess.equals(goldSet)) {
            numCorrect++;
          }
        }

      } else if (part == 2) {   // TF-IDF test
        System.out.println("TF-IDF Test");

        String[] queries = problem.split("; ");
        String[] golds = soln.split(", ");
        for (int i = 0; i < queries.length; i++) {
          numTotal++;

          String[] query = queries[i].split(", ");
          double guess = irSys.getTFIDFUnstemmed(query[0],
              new Integer(query[1]).intValue());

          double gold = new Double(golds[i]).doubleValue();
          if (guess >= gold - epsilon && guess <= gold + epsilon) {
            numCorrect++;
          }
        }
      } else if (part == 3) {
        System.out.println("Cosine Similarity Test");

        String[] queries = problem.split(", ");
        String[] golds = soln.split("; ");
        for (int i = 0; i < queries.length; i++) {
          numTotal++;

          PriorityQueue<Integer> guess = irSys.queryRank(queries[i]);
          double score = guess.getPriority();
          Integer docId = guess.next();

          String[] topGold = golds[i].split(", ");
          Integer topGoldId = new Integer(topGold[0]);
          double topScore = new Double(topGold[1]).doubleValue();
          
          if (docId.intValue() == topGoldId.intValue() &&
              score >= topScore - epsilon && score <= topScore + epsilon) {
            numCorrect++;
          }
        }
      }

      String feedback = numCorrect + "/" + numTotal + " Correct. " +
        "Accuracy: " + (double)numCorrect / numTotal;
      int points = 0;
      if (numCorrect == numTotal) points = 3;
      else if (numCorrect > 0.75 * numTotal) points = 2;
      else if (numCorrect > 0) points = 1;
      else points = 0;

      System.out.println("    Score: " + points + " Feedback: " + feedback);
    }

  }


  public static void main(String[] args) {
    String dataDir;

    IRSystem irSys = new IRSystem("../data/RiderHaggard");
    irSys.index();
    irSys.computeTFIDF();

    if (args.length == 0) {
      runTests(irSys);
    } else {
      String query = "";
      boolean haveAdded = false;
      for (String s : args) {
        if (haveAdded) {
          query += " ";
        }
        query += s;
        haveAdded = true;
      }
      PriorityQueue<Integer> results = irSys.queryRank(query);
      System.out.println("Best matching documents to '" + query + "':");
      int numResults = results.size();
      for (int i = 0; i < numResults; i++) {
        double score = results.getPriority();
        String title = irSys.getTitle(results.next().intValue());
        System.out.println(title + ": " + score);
      }
    }
  }
}
